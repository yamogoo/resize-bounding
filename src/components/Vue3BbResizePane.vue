<template>
  <div
    :class="['bb-resize__pane', `--${direction}`]"
    data-testid="bb-resize-pane"
  >
    <div
      ref="refPane"
      data-testid="bb-resize-pane-splitter"
      :class="['bb-resize__pane__splitter', isFocused ? 'show' : 'hide']"
    >
      <div
        v-if="isFocused || constantlyShowKnob"
        class="bb-resize__pane__splitter__icon"
        :style="knobStyle"
      ></div>
    </div>
  </div>
</template>

<script setup lang="ts">
import { onMounted, onUnmounted, ref, type Ref, computed } from "vue";

import type { BBResizeStyles, BBResizeOptions } from "./typings";

const props = withDefaults(defineProps<Props>(), {
  direction: PaneDirections.RIGHT,
  constantlyShowKnob: false,
});

const emits = defineEmits<{
  (e: Emits.DRAG_START, data: PaneEmittedData): void;
  (e: Emits.DRAG_MOVE, data: PaneEmittedData): void;
  (e: Emits.DRAG_END, data: PaneEmittedData): void;
  (e: Emits.FOCUS, isFocused: boolean): void;
}>();

const refPane: Ref<HTMLDivElement | null> = ref(null);

const isFocused = ref(false);
let isResizing = false;

const knobStyle = computed(() => {
  const isHorizontal = checkIsHorizontal();
  return `transform: rotate(${isHorizontal ? 0 : 90}deg);`;
});

const checkIsHorizontal = (): boolean => /[l | r | h]/.test(props.direction);
const checkIsVertical = (): boolean => /[t | b | v]/.test(props.direction);

const updateCursor = (state: boolean) => {
  let cursorActive: string;

  if (checkIsHorizontal()) {
    cursorActive = props.styles?.cursor?.active.horizontal ?? "col-resize";
  } else if (checkIsVertical()) {
    cursorActive = props.styles?.cursor?.active.vertical ?? "row-resize";
  } else cursorActive = "auto";

  if (refPane.value) refPane.value.style.cursor = state ? cursorActive : "auto";
};

const emitFocus = (state: boolean) => {
  emits(Emits.FOCUS, state);
};

const onFocus = (e: PointerEvent, isFocused: boolean): void => {
  e.stopPropagation();
  onSelected(isFocused);
};

const onSelected = (state: boolean): void => {
  isFocused.value = state;
  updateCursor(state);
  emitFocus(state);
};

const onDragStart = (e: PointerEvent): void => {
  e.preventDefault();
  e.stopImmediatePropagation();

  isResizing = true;
  onSelected(isResizing);

  const el = e.currentTarget as HTMLDivElement;
  el.setPointerCapture(e.pointerId);

  emits(Emits.DRAG_START, {
    x: Math.round(e.clientX),
    y: Math.round(e.clientY),
    dir: props.direction,
  });

  const onDragMove = (e: PointerEvent): void => {
    emits(Emits.DRAG_MOVE, {
      x: Math.round(e.clientX),
      y: Math.round(e.clientY),
      dir: props.direction,
    });
  };

  const onDragEnd = (e: PointerEvent): void => {
    isResizing = false;

    onSelected(isResizing);
    updateCursor(isResizing);

    const el = e.currentTarget as HTMLDivElement;
    el.releasePointerCapture(e.pointerId);

    el.removeEventListener("pointermove", onDragMove);
    el.removeEventListener("pointerup", onDragEnd);

    emits(Emits.DRAG_END, {
      x: Math.round(e.clientX),
      y: Math.round(e.clientY),
      dir: props.direction,
    });
  };

  el.addEventListener("pointermove", onDragMove);
  el.addEventListener("pointerup", onDragEnd);
};

const onDragCancel = (e: PointerEvent): void => {
  const el = e.currentTarget as HTMLDivElement;

  isResizing = false;
  el.releasePointerCapture(e.pointerId);

  emits(Emits.DRAG_END, {
    x: Math.round(e.clientX),
    y: Math.round(e.clientY),
    dir: props.direction,
  });
};

const addEventListeners = () => {
  const el = refPane.value;

  if (el) {
    el.addEventListener("pointerenter", (e) => onFocus(e, true));
    el.addEventListener("pointerleave", (e) => onFocus(e, false));

    el.addEventListener("pointerdown", onDragStart);
    el.addEventListener("pointercancel", onDragCancel);
  }
};

const removeEventListeners = () => {
  const el = refPane.value;

  if (el) {
    el.removeEventListener("pointerenter", (e) => onFocus(e, true));
    el.removeEventListener("pointerleave", (e) => onFocus(e, false));

    el.removeEventListener("pointerdown", onDragStart);
    el.addEventListener("pointercancel", onDragCancel);
  }
};

onMounted(addEventListeners);
onUnmounted(removeEventListeners);

defineExpose({ refPane });
</script>

<script lang="ts">
export interface Props {
  direction?: PaneDirectionKey;
  constantlyShowKnob?: boolean;
  styles?: Partial<BBResizeStyles>;
  options?: Partial<BBResizeOptions>;
}

export enum PaneDirectionAliases {
  HORIZONTAL = "h",
  VERTICAL = "v",
}

export enum PaneDirections {
  LEFT = "l",
  RIGHT = "r",
  BOTTOM = "b",
  TOP = "t",
}

export type PaneDirectionKey = PaneDirections | PaneDirections | string;

export enum Emits {
  FOCUS = "focus",
  DRAG_START = "drag:start",
  DRAG_MOVE = "drag:move",
  DRAG_END = "drag:end",
}

export interface PaneEmittedData {
  x: number;
  y: number;
  dir: string;
}
</script>

<style lang="scss">
$__show-color: blue;
$__pressed-color: #3655e171;

.bb-resize {
  .bb-resize__pane__splitter {
    display: flex;
    align-items: center;
    justify-content: center;
    z-index: 9999;
  }

  .bb-resize__pane__splitter__icon {
    position: absolute;
    width: 12px;
    height: 120px;
    border-radius: 12px;
    background-color: red;
    margin: auto;
  }

  &__pane {
    position: absolute;
    display: block;
    z-index: 9999;

    &.--l {
      left: 2px;
      width: 0px;
      height: 100%;

      .bb-resize__pane__splitter {
        right: 0;
        width: 4px;
        height: 100%;
      }
    }

    &.--r {
      right: 2px;
      width: 0px;
      height: 100%;

      .bb-resize__pane__splitter {
        left: 0;
        width: 4px;
        height: 100%;
      }
    }

    &.--b {
      bottom: -2px;
      height: 0px;
      width: 100%;

      .bb-resize__pane__splitter {
        top: 0;
        width: 100%;
        height: 4px;
      }
    }

    &.--t {
      top: 2px;
      height: 0px;
      width: 100%;

      .bb-resize__pane__splitter {
        bottom: 0;
        width: 100%;
        height: 4px;
      }
    }

    &__splitter {
      position: absolute;
      display: block;
      z-index: 0;

      &.show {
        background-color: $__show-color;
      }

      &.pressed {
        background-color: $__pressed-color;
      }
    }
  }
}
</style>
