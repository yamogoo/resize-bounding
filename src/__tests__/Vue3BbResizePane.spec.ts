import { describe, test, expect } from "vitest";
import { shallowMount, VueWrapper } from "@vue/test-utils";

import { DataTestIds } from "./setup";

import Vue3BbResizePane, {
  Emits,
  PaneDirectionAliases,
  PaneDirections,
  type PaneEmittedData,
  type Props,
} from "@/components/Vue3BbResizePane.vue";

describe("Vue3BbResize", () => {
  beforeEach(() => {
    Element.prototype.setPointerCapture = vi.fn();
    Element.prototype.releasePointerCapture = vi.fn();
  });

  describe("classes", () => {
    test('should have "hide class by default"', () => {
      const wrapper = shallowMount(Vue3BbResizePane);

      const splitterEl = wrapper.find(
        `[data-testid="${DataTestIds.BB_RESIZE_PANE_SPLITTER}"]`,
      );

      expect(splitterEl.classes("hide")).toBeTruthy();
      expect(splitterEl.classes()).toMatchSnapshot();
    });
  });
  describe("events", () => {
    describe("focus", async () => {
      const wrapper = shallowMount(Vue3BbResizePane, {
        props: {
          direction: PaneDirections.RIGHT,
        },
      });

      const rootEl = wrapper.find(
        `[data-testid="${DataTestIds.BB_RESIZE_PANE_SPLITTER}"]`,
      );

      test(`should emit "${Emits.FOCUS}" with state (true)`, async () => {
        await rootEl.trigger("pointerenter");
        const isFocused: boolean = (
          wrapper.emitted(Emits.FOCUS) as boolean[][]
        )[0][0];

        expect(isFocused).toBeTruthy();
        expect(isFocused).toMatchSnapshot();
      });

      test(`should emit "${Emits.FOCUS}" with state (false)`, async () => {
        await rootEl.trigger("pointerleave");
        const isFocused = (wrapper.emitted(Emits.FOCUS) as boolean[][])[1][0];

        expect(isFocused).toBeFalsy();
        expect(isFocused).toMatchSnapshot();
      });
    });

    describe("drag", () => {
      const wrapper = shallowMount(Vue3BbResizePane, {
        props: {
          direction: PaneDirections.RIGHT,
        },
      });

      const rootEl = wrapper.find(
        `[data-testid="${DataTestIds.BB_RESIZE_PANE_SPLITTER}"]`,
      );

      test(`should emit "${Emits.DRAG_START} -> ${Emits.DRAG_MOVE} -> ${Emits.DRAG_END}" with {x, y, dir}`, async () => {
        const checkData = (
          eventName: string,
          wrapper: VueWrapper,
          eventData: { clientX: number; clientY: number },
        ): PaneEmittedData => {
          const ev = wrapper.emitted(eventName) as PaneEmittedData[][];
          const data = ev[0][0];
          const { x, y } = data;

          const { clientX, clientY } = eventData;

          expect(data).toHaveProperty("x");
          expect(x).toBe(clientX);
          expect(data).toHaveProperty("y");
          expect(y).toBe(clientY);
          expect(data).toHaveProperty("dir");
          expect(data.dir).toBe(PaneDirections.RIGHT);
          return data;
        };

        let events: PaneEmittedData[] = [];

        const makeTrigger = async (
          event: string,
          eventData: { clientX: number; clientY: number },
          emittedEventName: string,
        ) => {
          await rootEl.trigger(event, eventData);
          events.push(checkData(emittedEventName, wrapper, eventData));
        };

        await rootEl.trigger("pointerenter", { clientX: 1, clientY: 1 });
        await makeTrigger(
          "pointerdown",
          { clientX: 1, clientY: 1 },
          Emits.DRAG_START,
        );
        await makeTrigger(
          "pointermove",
          { clientX: 2, clientY: 2 },
          Emits.DRAG_MOVE,
        );
        await makeTrigger(
          "pointerup",
          { clientX: 3, clientY: 3 },
          Emits.DRAG_END,
        );

        expect(events.length).toBe(3);
        expect(events).toMatchSnapshot();
      });
    });
  });

  describe("styles", () => {
    test('default cursor style should be "auto"', async () => {
      const wrapper = shallowMount(Vue3BbResizePane);
      await wrapper.setProps({ direction: PaneDirections.RIGHT });

      const rootEl = wrapper.find(
        `[data-testid="${DataTestIds.BB_RESIZE_PANE_SPLITTER}"]`,
      );

      await rootEl.trigger("pointerenter");
      await rootEl.trigger("pointerleave");

      const cursorStyle = rootEl.attributes("style");

      expect(cursorStyle).toContain("auto");
      expect(cursorStyle).toMatchSnapshot();
    });

    const checkCursorActive = async (
      wrapper: VueWrapper,
      props: Props,
      comparedCursorStyle: string,
    ): Promise<void> => {
      await wrapper.setProps(props);

      const rootEl = wrapper.find(
        `[data-testid="${DataTestIds.BB_RESIZE_PANE_SPLITTER}"]`,
      );
      await rootEl.trigger("pointerenter");

      const cursorStyle = rootEl.attributes("style");

      expect(cursorStyle).toContain(comparedCursorStyle);
      expect(cursorStyle).toMatchSnapshot();
    };

    describe("active", () => {
      describe("default styles", () => {
        test.each([
          PaneDirectionAliases.HORIZONTAL,
          PaneDirections.RIGHT,
          PaneDirections.LEFT,
        ])(
          'default cursor style should be "col-resize" (--%s)',
          async (direction) => {
            const wrapper = shallowMount(Vue3BbResizePane);
            checkCursorActive(wrapper, { direction }, "col-resize");
          },
        );

        test.each([
          PaneDirectionAliases.VERTICAL,
          PaneDirections.TOP,
          PaneDirections.BOTTOM,
        ])(
          'default cursor style should be "row-resize" (--%s)',
          async (direction) => {
            const wrapper = shallowMount(Vue3BbResizePane);
            checkCursorActive(wrapper, { direction }, "row-resize");
          },
        );
      });

      describe("custom styles", () => {
        const CUSTOM_ACTIVE_CURSOR = "ew-resize";

        test.each([...Object.values(PaneDirections)])(
          `custom cursor style should be "${CUSTOM_ACTIVE_CURSOR}" (--%s)`,
          async (direction) => {
            const wrapper = shallowMount(Vue3BbResizePane, {
              props: {
                styles: {
                  cursor: {
                    active: {
                      vertical: CUSTOM_ACTIVE_CURSOR,
                      horizontal: CUSTOM_ACTIVE_CURSOR,
                    },
                  },
                },
              },
            });
            checkCursorActive(
              wrapper,
              {
                direction,
              },
              CUSTOM_ACTIVE_CURSOR,
            );
          },
        );
      });
    });
  });
});
