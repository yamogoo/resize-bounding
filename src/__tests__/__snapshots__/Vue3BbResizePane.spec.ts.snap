// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Vue3BbResize > classes > should have "hide class by default" 1`] = `
[
  "bb-resize__pane__splitter",
  "hide",
]
`;

exports[`Vue3BbResize > events > drag > should emit "drag:start -> drag:move -> drag:end" with {x, y, dir} 1`] = `
[
  {
    "dir": "r",
    "x": 1,
    "y": 1,
  },
  {
    "dir": "r",
    "x": 2,
    "y": 2,
  },
  {
    "dir": "r",
    "x": 3,
    "y": 3,
  },
]
`;

exports[`Vue3BbResize > events > focus > should emit "focus" with state (false) 1`] = `false`;

exports[`Vue3BbResize > events > focus > should emit "focus" with state (true) 1`] = `true`;

exports[`Vue3BbResize > styles > active > custom styles > custom cursor style should be "ew-resize" (--b) 1`] = `"cursor: ew-resize;"`;

exports[`Vue3BbResize > styles > active > custom styles > custom cursor style should be "ew-resize" (--l) 1`] = `"cursor: ew-resize;"`;

exports[`Vue3BbResize > styles > active > custom styles > custom cursor style should be "ew-resize" (--r) 1`] = `"cursor: ew-resize;"`;

exports[`Vue3BbResize > styles > active > custom styles > custom cursor style should be "ew-resize" (--t) 1`] = `"cursor: ew-resize;"`;

exports[`Vue3BbResize > styles > active > default styles > default cursor style should be "col-resize" (--h) 1`] = `"cursor: col-resize;"`;

exports[`Vue3BbResize > styles > active > default styles > default cursor style should be "col-resize" (--l) 1`] = `"cursor: col-resize;"`;

exports[`Vue3BbResize > styles > active > default styles > default cursor style should be "col-resize" (--r) 1`] = `"cursor: col-resize;"`;

exports[`Vue3BbResize > styles > active > default styles > default cursor style should be "row-resize" (--b) 1`] = `"cursor: row-resize;"`;

exports[`Vue3BbResize > styles > active > default styles > default cursor style should be "row-resize" (--t) 1`] = `"cursor: row-resize;"`;

exports[`Vue3BbResize > styles > active > default styles > default cursor style should be "row-resize" (--v) 1`] = `"cursor: row-resize;"`;

exports[`Vue3BbResize > styles > default cursor style should be "auto" 1`] = `"cursor: auto;"`;
